core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -623,37 +683,114 @@ class LogManager(logDirs: Array[File],
+        if (!destLog.renameDir(Log.logDirName(topicPartition))) {
+          logDirFailureChannel.maybeAddOfflineLogDir(destLog.dir.getParent, s"Failed to rename dir for partition $topicPartition",
+            new IOException(s"Failed to rename dir for partition $topicPartition"))
+          // Revert the source log directory name so that it will not be deleted
+          if (sourceLog.renameDir(Log.logDirName(topicPartition))) {
+            sourceRenamedForDelete = false
+            destRenamedForSwap = false
+            errorMsgOpt = Some(s"Failed to rename destination log directory ${destLog.dir.getAbsolutePath} to be the primary partition")
+          } else {
+            sourceRenamedForDelete = true
+            destRenamedForSwap = false
+            logDirFailureChannel.maybeAddOfflineLogDir(sourceLog.dir.getParent, s"Failed to rename dir for partition $topicPartition",
+              new IOException(s"Failed to rename dir for partition $topicPartition"))
+            errorMsgOpt = Some(s"Failed to rename destination log directory ${destLog.dir.getAbsolutePath} to " +
+              s"be the primary partition and failed to rename source log directory ${sourceLog.dir.getAbsolutePath} to avoid deletion")
+          }
+        }
+      } else {
+        logDirFailureChannel.maybeAddOfflineLogDir(sourceLog.dir.getParent, s"Failed to rename dir for partition $topicPartition",
+          new IOException(s"Failed to rename dir for partition $topicPartition"))
+        sourceRenamedForDelete = false
+        destRenamedForSwap = false
+        errorMsgOpt = Some(s"Failed to rename source log directory ${sourceLog.dir.getAbsolutePath} for deletion")
+      }
+
+      /*
+       * It is possible for the log of this partition to be mistakenly deleted from this broker if the following sequence
+       * of events happen. However, the chance of this happening is extremely low. And if the replication factor > 1, the broker
+       * should be able to re-fetch the log for this partition from the leader broker if the log is deleted by mistake.
+       *
+       * 1) Both source log directory and destination log directory are online and the destination log catches up with the source log
+       * 2) Source log is successfully renamed for deletion
+       * 3) Destination log can not be renamed to be the primary log due to failure in the destination log directory
+       * 4) Source log can not be renamed to skip deletion due to failure in the source log directory.
+       * 5) After broker restarts, the source log directory is online but the destination log directory is offline.
+       *    Then source log is deleted because it is marked for deletion.
+       * 6) Broker restarts again, both source log directory and destination log directory are online.
+       *    Then the destination log is deleted because all log directories are online and there is no primary log for
+       *    this partition on the broker.
```

Comment:
My gut feel is that the logic is harder to be correct if we add more suffix and more rename steps.

For example, the following sequence of events may happen with the suggested approach:

- FutureLog is successfully renamed to log.completed.
- CurrentLog can not be renamed due to log directory failure.
- log.completed is renamed to log. If we do not rename it to log now, then when the broker starts with CurrentLog still offfline, broker will forget about CurrentLog and will need to rename log.completed to log
- Now broker restarts again with CurrentLog online. Broker sees two copies of log for the same partition. It will be hard for broker to decide which one to use if the log has been truncated when CurrentLog is still offline.

I think the current approach has simpler with less suffix. It will cause the replica to be deleted in very rare scenario. And even then the broker should be able to recover from this as long as RF > 1. If RF = 1, Kafka can not ensure no data loss anyway due to disk failure.

Reply:
It seems that all other issues have been addressed. I need to think more about this issue more. Let me first fix the test, cleanup the patch and upload it.

Reply:
@junrao I think the suggested approach will work. I am just wondering if the following approach would be simpler:

In LogManager.deleteAndSwapFutureLog():

1. Rename FutureLog to be the current log
2. Rename CurrentLog to be deleted.

And here is how we handle log directory failure:

1) FutureLog is successfully renamed to be the current log
2) CurrentLog can not be renamed due to log directory failure
3) The log in the source log directory will be marked offline. The log in the destination log directory will serve as the current log and this partition is still online.
4) Broker is restarted and the source log directory is still offline. Nothing needs to be done. The partition is online.
5) Broker is restarted and the source log directory is online. But the destination log directory is online. In this case the log in the source log directory will be truncated based on the leader epoch and will be try to catch up with the leader.
6) Broker is restarted and both source and destination log directory is offline. LogManager.loadLogs() will choose the one with the larger latestEpoch as the current log. If these two replicas has the same leader epoch, then the one with the larger nextOffset will be chosen as the current log. The other log will be marked for deletion.

The advantage of this solution is that we don't need to new suffix, and we don't need to have the logic of renaming a log directory back-and-forth. This solution takes advantage of the leaderEpoch to resolve conflicts. Do you think this would be a good solution?

Reply:
@lindong28 : What you suggested is simpler. My concerns are the following: (1) It can't truly protect the case when the same partition shows up in more than one log dir (say by human mistakes). (2) The approach of using LeaderEpoch only works when the message format has been upgraded. However, users may not upgrade the message format immediately after upgrading the code. Picking the replica based on just the offset is less reliable.

Reply:
@junrao Regarding (1), if the same partition appears on multiple log directories due to human either, I think log manager can still choose the one with the highest epoch (or nextOffset if epoch is the same) to address the problem. Is my understanding right?

Regarding (2), I am wondering if it is reasonable to support intra-broker replica movement only if message format has been increased to support leader epoch. I personally think it is a good tradeoff to keep the Kafka implementation simpler in the long term. It seems reasonable for a new feature to rely on a message format that has come before it.

BTW, in my solution to another issue you raised, i.e. inconsistency between the current and the future replica, I also used this trick of leader epoch to address the problem. The solution to that problem will probably be less clean if we can not reply on leader epoch. I will need to think more about how to address these two problems if we want to allow user to move replica within broker with older message format.

Reply:
Also note that KIP-112 is only enabled if inter-broker-protocol is 1.0 or higher.

Reply:
@lindong28: For (1), if there is human error, the 2 replicas could correspond to the same topic created at different times. So, their leader epoch may not be directly comparable. It just feels safer if we have a more direct way to detect errors like this. For (2), note that inter-broker-protocol is separate from the message format. Even when the message format is set, leader epoch only applies to newly produced messages.

Reply:
@junrao Regarding (2), I am thinking that it may be reasonable to say we only support inter-broker replica movement only if message format is supports epoch. This is similar to saying that we only support time-based query if message format supports time index, and we only KafkaHeader if the message format support Kafka Header. This may be worth doing if it can keep the Kafka implementation simpler in the long term.

Also regarding (2), my understanding is that we can use leader epoch to choose the newer log directory in this case after message format has been upgraded to support leader epoch, even if there is still old messages in the log that doesn't have leader epoch. This is because the trick here only look at the epoch of the latest message for each log directory. If this is not true then my approach would not work.

Regarding (1), I realized the following shortcomings of the suggested approach while trying to implement it. Let me compare the suggested approach with my approach in more detail below.

To clarify, my approach is the following:
1) Rename FutureLog to be the current log
2) Rename CurrentLog to be deleted.

The approach you suggested is the following:
1) Rename FutureLog to log.completed.
2) Rename CurrentLog to log.deleted.
3) Rename log.completed to log.

Pros of the suggested approach:
- It allows user to move replica between log directories of the same broker before user has upgraded to the message format that support leader epoch. (a short term benefit)
- It matches how we do swapping in log cleaner (I am not very sure the benefit of following the existing swapping logic in log cleaner in this case though. Can you explain a bit more?)
- It can help us detect human error if the user mistakenly copied a directory with old data to a broker which already has directory for this partition on another log directory.

Cons of the suggested approach:
- If we fail at the step 2 (Rename CurrentLog to log.deleted), we will mark the partition offline even if we can use the partition in the destination log directory. This reduces availability of the partition.
- If we fail at step 3 (Rename log.completed to log), we will still delete the replica in the source log directory because it has been successfully renamed for deletion. This reduces the persistence of the partition.
- Extra logic is needed to keep completable logs in a separate map in LogManager so that we can include its size in the DescribeLogDirsResponse. We also need extra metric tag for Log whose directory has be renamed to log.completed. This adds complexity to Kafka implementation in the long term.
- It doesn't protect against more complicated errors, e.g. if user deletes new log directory before copying the old log directory, or if user deletes specific segment for the partition. 

Overall I think the ability to detect some human error is nice to have. But human error detection is generally hard to do and we don't have a good definition for the scope of human error that can be detected. I am not sure if this is worth the cost in availability and code complexity. Actually, if we do want to detect this specific human error, we can also modify the my approach to have broker refuse to start if the same partition appears in more than one directory. This will generate false positive only if source log directory fails exactly at the time the destination partition has caught up with the source partition, which should be pretty. Since user knows the log directory that was offline, he/she can manually delete the partition in the source log directory correctly and restart the broker.

The reply is a bit long. Thanks for taking time to read and discuss this issue.

Reply:
@lindong28 : I like the modified version of your approach. Basically fail the broker if we detect duplicated logs during restart. This makes the logic simpler. With this, we probably don't need to have the constraint on message format.

Reply:
@junrao @ijuma Great! I will implement this version in the updated patch.